src/sinol_make/commands/run/__init__.py

```diff
@@ -368,70 +379,153 @@ def print_view(output_file=None):
+		for program in expected_scores.keys():
+			print("  %s:" % program)
+			print("    points: %d" % expected_scores[program]["points"])
+			print("    expected: " + str(expected_scores[program]["expected"]))
+
+
+	def validate_expected_scores(self, results, programs):
+		new_expected_scores = {} # Expected scores based on results
+
+		for program in results.keys():
+			new_expected_scores[program] = {
+				"expected": results[program],
+				"points": self.calculate_points(results[program])
+			}
+
+		expected_scores = {} # Expected scores from config with only programs and groups that were run
+		if "sinol_expected_scores" in self.config:
+			for program in results.keys():
+				if program in self.config["sinol_expected_scores"]:
+					expected_scores[program] = {
+						"expected": {},
+						"points": 0
+					}
+
+					for group in results[program].keys():
+						if group in self.config["sinol_expected_scores"][program]["expected"]:
+							expected_scores[program]["expected"][group] = self.config["sinol_expected_scores"][program]["expected"][group]
+
+					expected_scores[program]["points"] = self.calculate_points(expected_scores[program]["expected"])
+
+		print(util.bold("Expected scores from config:"))
+		self.print_expected_scores(expected_scores)
+		print(util.bold("\nExpected scores based on results:"))
+		self.print_expected_scores(new_expected_scores)
+
+		expected_scores_diff = dictdiffer.diff(expected_scores, new_expected_scores)
+		added_programs = set()
+		removed_programs = set()
+		added_groups = set()
+		removed_groups = set()
+
+		for type, field, change in list(expected_scores_diff):
+			if type == "add":
+				if field == '': # Programs were added
+					for program in change:
+						added_programs.add(program[0])
+				elif field[1] == "expected": # Groups were added
+					for group in change:
+						added_groups.add(group[0])
+			elif type == "change":
+				if field[1] == "expected": # Result for group changed
+					program = field[0]
+					group = field[2]
+					old_result = change[0]
+					group = change[1]
+
+					print(util.warning("Program %s passed group %d with status %s while it should pass with status %s." %
+										(program, group, group, old_result)))
+
+		# Only if sinol_make was run on all programs we should check for removed programs
+		if self.args.programs == None and "sinol_expected_scores" in self.config:
+			for program in self.config["sinol_expected_scores"].keys():
+				if program not in new_expected_scores.keys():
+					removed_programs.add(program)
+
+		# Only if sinol_make was run on all groups we should check for removed groups
+		if self.args.tests == None and "sinol_expected_scores" in self.config:
+			for program in self.config["sinol_expected_scores"].keys():
+				for group in self.config["sinol_expected_scores"][program]["expected"].keys():
+					if program in new_expected_scores and group not in new_expected_scores[program]["expected"].keys():
+						removed_groups.add(group)
```

Comment:
The removed programs here were created differently than the added programs. It seems inconsistent and there is more surface for bugs to appear. Was there a reason why they need to be created in other ways?

Reply:
Now removed programs and groups are checked with `dictdiffer`. 954c7b27189063d343f5a1d075992c9f5a185e67